packages/react-router/lib/hooks.tsx

```diff
@@ -559,7 +560,7 @@ export function _renderMatches(
     if (dataRouterState?.errors) {
       // Don't bail if we have data router errors so we can render them in the
       // boundary.  Use the pre-matched (or shimmed) matches
-      matches = dataRouterState.matches;
+      matches = dataRouterState.matches as DataRouteMatch[];
```

Comment:
Since we're inside a data router initialized with react-specific routes, we can safely cast these back here when we pull them from the agnostic router state.